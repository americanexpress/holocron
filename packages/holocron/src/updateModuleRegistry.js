/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import shallowEqual from 'shallowequal';
import {
  getModules, getModuleMap, resetModuleRegistry, addHigherOrderComponent,
} from './moduleRegistry';
import loadModule from './loadModule.node';

export function areModuleEntriesEqual(firstModuleEntry, secondModuleEntry) {
  return !!(
    firstModuleEntry
    && secondModuleEntry
    && shallowEqual(firstModuleEntry.browser, secondModuleEntry.browser)
    && shallowEqual(firstModuleEntry.legacyBrowser, secondModuleEntry.legacyBrowser)
    && shallowEqual(firstModuleEntry.node, secondModuleEntry.node)
  );
}

function defaultGetModulesToUpdate(curr, next) {
  return Object
    .keys(next)
    .filter((moduleName) => !areModuleEntriesEqual(curr[moduleName], next[moduleName]));
}

export default async function updateModuleRegistry({
  moduleMap: nextModuleMap,
  onModuleLoad = () => null,
  batchModulesToUpdate = (x) => [x],
  getModulesToUpdate = defaultGetModulesToUpdate,
  listRejectedModules,
}) {
  const currentModuleMap = getModuleMap().toJS();
  const modulesToUpdate = batchModulesToUpdate(
    getModulesToUpdate(currentModuleMap.modules || {}, nextModuleMap.modules)
  );
  const flatModulesToUpdate = modulesToUpdate.reduce((acc, batch) => [...acc, ...batch], []);
  const rejectedModules = {};
  let successfullyLoadedModules = await modulesToUpdate.reduce(async (acc, moduleBatch) => {
    const previouslyResolvedModules = await acc;
    const newlyResolvedModules = await Promise.allSettled(moduleBatch.map(
      async (moduleName) => {
        try {
          const loadedModule = await loadModule(
            moduleName,
            nextModuleMap.modules[moduleName],
            onModuleLoad
          );
          return addHigherOrderComponent(loadedModule);
        } catch (e) {
          const brokenUrl = nextModuleMap.modules[moduleName].node.url;
          if (currentModuleMap.modules && currentModuleMap.modules[moduleName]) {
            const previousUrl = currentModuleMap.modules[moduleName].node.url;
            // eslint-disable-next-line no-console
            console.error(`There was an error loading module ${moduleName} at ${brokenUrl}. Reverting back to ${previousUrl}`, e);
          } else {
            // eslint-disable-next-line no-console
            console.error(`There was an error loading module ${moduleName} at ${brokenUrl}. Ignoring ${moduleName} until next module map poll.`, e);
          }
          rejectedModules[moduleName] = {
            ...nextModuleMap.modules[moduleName],
            reasonForRejection: e.message,
          };
          return Promise.reject(e);
        }
      }
    ));
    const fulfilledModules = newlyResolvedModules.filter(
      ({ status }) => status === 'fulfilled'
    ).map(({ value }) => value);
    return [...previouslyResolvedModules, ...fulfilledModules];
  }, []);
  const rejectedModuleNames = Object.keys(rejectedModules);
  const updatedFlatMap = flatModulesToUpdate.filter((mod) => !rejectedModuleNames.includes(mod));
  successfullyLoadedModules = successfullyLoadedModules.reduce(
    (acc, module, i) => ({
      ...acc,
      [updatedFlatMap[i]]: module,
    }),
    {}
  );
  const newModules = getModules().merge(successfullyLoadedModules);
  const nextModules = { ...nextModuleMap.modules };
  // Keep working version of modules if they are in the list of problem modules
  rejectedModuleNames.forEach((module) => {
    if (currentModuleMap.modules) {
      nextModules[module] = currentModuleMap.modules[module];
    } else {
      // If it doesn't exist in the old module map, that means its being deployed for the first time
      // or holocron is initializing modules for the first time. If there is an issue with the
      // module we should exclude it from the module map entirely.
      delete nextModules[module];
    }
  });
  resetModuleRegistry(newModules, { ...nextModuleMap, modules: nextModules });
  const loadedModules = updatedFlatMap.reduce(
    (acc, moduleName) => ({ ...acc, [moduleName]: nextModules[moduleName] }),
    {}
  );

  if (listRejectedModules) {
    return { loadedModules, rejectedModules };
  }
  return loadedModules;
}
