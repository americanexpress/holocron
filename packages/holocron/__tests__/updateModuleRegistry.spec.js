/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import updateModuleRegistry, { areModuleEntriesEqual } from '../src/updateModuleRegistry';
import loadModule from '../src/loadModule.node';
import {
  getModules,
  getModuleMap,
  resetModuleRegistry,
  addHigherOrderComponent,
} from '../src/moduleRegistry';

function AwesomeModule() {
  return 'initial awesome-module';
}
function AnotherModule() {
  return 'initial awesome-module';
}
function ModuleThree() {
  return 'initial awesome-module';
}

jest.mock('../src/loadModule.node', () => {
  const createTimeoutPromise = (
    resolveWith,
    time
  ) => new Promise(
    (res) => setTimeout(() => res(resolveWith), time)
  );

  const makeUpdatedModule = (moduleName, moduleVersion) => function UpdatedModule() {
    return `new ${moduleName}@${moduleVersion}`;
  };

  return jest.fn((moduleName, moduleVersion) => createTimeoutPromise(
    makeUpdatedModule(moduleName, moduleVersion),
    100
  ));
});

describe('updateModuleRegistry', () => {
  const onModuleLoad = jest.fn();

  const getModuleOutputs = (modules) => modules.reduce((acc, module, moduleName) => ({
    ...acc,
    [moduleName]: module,
  }), {});

  beforeEach(() => {
    jest.clearAllMocks();
    resetModuleRegistry(
      {
        'awesome-module': addHigherOrderComponent(AwesomeModule),
        'another-module': addHigherOrderComponent(AnotherModule),
        'module-three': addHigherOrderComponent(ModuleThree),
      },
      {
        key: 'key123',
        modules: {
          'awesome-module': {
            node: {
              url: 'https://example.com/cdn/awesome-module/1.0.0/awesome-module.node.js',
              integrity: '47474',
            },
            browser: {
              url: 'https://example.com/cdn/awesome-module/1.0.0/awesome-module.browser.js',
              integrity: '465345v',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/awesome-module/1.0.0/awesome-module.legacy.browser.js',
              integrity: '4564632',
            },
          },
          'another-module': {
            node: {
              url: 'https://example.com/cdn/another-module/1.0.0/another-module.node.js',
              integrity: '6786784',
            },
            browser: {
              url: 'https://example.com/cdn/another-module/1.0.0/another-module.browser.js',
              integrity: '66465879',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/another-module/1.0.0/another-module.legacy.browser.js',
              integrity: '4776754',
            },
          },
          'module-three': {
            node: {
              url: 'https://example.com/cdn/module-three/1.0.0/module-three.node.js',
              integrity: '7747545',
            },
            browser: {
              url: 'https://example.com/cdn/module-three/1.0.0/module-three.browser.js',
              integrity: '44532345',
            },
            legacyBrowser: {
              url: 'https://example.com/cdn/module-three/1.0.0/module-three.legacy.browser.js',
              integrity: '3534536',
            },
          },
        },
      }
    );
  });

  it('should only fetch the changed modules', async () => {
    expect.assertions(2);

    const updatedModuleData = {
      node: {
        url: 'https://example.com/cdn/another-module/2.4.0/another-module.node.js',
        integrity: '235534',
      },
      browser: {
        url: 'https://example.com/cdn/another-module/2.4.0/another-module.browser.js',
        integrity: '56643',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/another-module/2.4.0/another-module.legacy.browser.js',
        integrity: '1145',
      },
    };
    const currentModuleMap = getModuleMap().toJS();
    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': updatedModuleData,
      },
    };
    const batchModulesToUpdate = (x) => [x];
    await updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
      batchModulesToUpdate,
    });
    expect(loadModule).toHaveBeenCalledTimes(1);
    expect(loadModule).toHaveBeenCalledWith('another-module', updatedModuleData, onModuleLoad);
  });

  it('should use a default onModuleLoad', async () => {
    expect.assertions(1);
    const updatedModuleData = {
      node: {
        url: 'https://example.com/cdn/another-module/2.4.0/another-module.node.js',
        integrity: '2342352',
      },
      browser: {
        url: 'https://example.com/cdn/another-module/2.4.0/another-module.browser.js',
        integrity: 'w445',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/another-module/2.4.0/another-module.legacy.browser.js',
        integrity: '3455c',
      },
    };
    const currentModuleMap = getModuleMap().toJS();
    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': updatedModuleData,
      },
    };

    await updateModuleRegistry({
      moduleMap: newModuleMap,
    });

    expect(loadModule.mock.calls[0][2]()).toBe(null);
  });

  it('should batch the requests', async () => {
    const currentModuleMap = getModuleMap().toJS();
    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': {
          node: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.node.js',
            integrity: '333354',
          },
          browser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.browser.js',
            integrity: '13456',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.legacy.browser.js',
            integrity: '4456643',
          },
        },
        'yet-another-module': {
          node: {
            url: 'https://example.com/cdn/yet-another-module/2.4.0/yet-another-module.node.js',
            integrity: '8975431',
          },
          browser: {
            url: 'https://example.com/cdn/yet-another-module/2.4.0/yet-another-module.browser.js',
            integrity: '23566',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/yet-another-module/2.4.0/yet-another-module.legacy.browser.js',
            integrity: '345346',
          },
        },
        'cool-module': {
          node: {
            url: 'https://example.com/cdn/cool-module/2.4.0/cool-module.node.js',
            integrity: '893312123',
          },
          browser: {
            url: 'https://example.com/cdn/cool-module/2.4.0/cool-module.browser.js',
            integrity: '225356',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/cool-module/2.4.0/cool-module.legacy.browser.js',
            integrity: '789685',
          },
        },
        'cooler-module': {
          node: {
            url: 'https://example.com/cdn/cooler-module/2.4.0/cooler-module.node.js',
            integrity: '8905454',
          },
          browser: {
            url: 'https://example.com/cdn/cooler-module/2.4.0/cooler-module.browser.js',
            integrity: '63478',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/cooler-module/2.4.0/cooler-module.legacy.browser.js',
            integrity: '35523',
          },
        },
      },
    };

    const batchModulesToUpdate = (moduleNames) => moduleNames.reduce((acc, moduleName) => {
      const lastArr = acc[acc.length - 1];
      if (lastArr.length < 2) lastArr.push(moduleName);
      else acc.push([moduleName]);
      return acc;
    }, [[]]);

    const startTime = performance.now();
    await updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
      batchModulesToUpdate,
    });
    const executionTime = performance.now() - startTime;

    expect(executionTime).toBeGreaterThanOrEqual(200);
    expect(executionTime).toBeLessThan(300);
  });

  it('should put the requests in a single batch by default', async () => {
    const currentModuleMap = getModuleMap().toJS();
    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': {
          node: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.node.js',
            integrity: '123',
          },
          browser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.browser.js',
            integrity: '234',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.legacy.browser.js',
            integrity: '344',
          },
        },
        'yet-another-module': {
          node: {
            url: 'https://example.com/cdn/yet-another-module/2.4.0/yet-another-module.node.js',
            integrity: '123',
          },
          browser: {
            url: 'https://example.com/cdn/yet-another-module/2.4.0/yet-another-module.browser.js',
            integrity: '234',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/yet-another-module/2.4.0/yet-another-module.legacy.browser.js',
            integrity: '344',
          },
        },
        'cool-module': {
          node: {
            url: 'https://example.com/cdn/cool-module/2.4.0/cool-module.node.js',
            integrity: '123',
          },
          browser: {
            url: 'https://example.com/cdn/cool-module/2.4.0/cool-module.browser.js',
            integrity: '234',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/cool-module/2.4.0/cool-module.legacy.browser.js',
            integrity: '344',
          },
        },
        'cooler-module': {
          node: {
            url: 'https://example.com/cdn/cooler-module/2.4.0/cooler-module.node.js',
            integrity: '123',
          },
          browser: {
            url: 'https://example.com/cdn/cooler-module/2.4.0/cooler-module.browser.js',
            integrity: '234',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/cooler-module/2.4.0/cooler-module.legacy.browser.js',
            integrity: '344',
          },
        },
      },
    };

    const startTime = performance.now();
    await updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
    });
    const executionTime = performance.now() - startTime;

    expect(executionTime).toBeGreaterThanOrEqual(100);
    expect(executionTime).toBeLessThan(200);
  });

  it('should update the modules in the registry', async () => {
    expect.assertions(2);
    const currentModuleMap = getModuleMap().toJS();
    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': {
          node: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.node.js',
            integrity: '123',
          },
          browser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.browser.js',
            integrity: '234',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.legacy.browser.js',
            integrity: '344',
          },
        },
      },
    };
    const batchModulesToUpdate = (x) => [x];

    await updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
      batchModulesToUpdate,
    });

    expect(getModuleOutputs(getModules())).toMatchSnapshot('modules');
    expect(getModuleMap()).toMatchSnapshot('module map');
  });

  it('should remove modules from the registry', async () => {
    expect.assertions(3);
    expect(getModuleMap().getIn(['modules', 'awesome-module'])).toMatchSnapshot();
    expect(getModules().getIn(['awesome-module', 'displayName'])).toBe('Connect(HolocronModule(AwesomeModule))');

    const newModuleMap = {
      key: 'key-123',
      modules: {
        'another-module': {
          node: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.node.js',
            integrity: '578996',
          },
          browser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.browser.js',
            integrity: '53344335',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/another-module/2.4.0/another-module.legacy.browser.js',
            integrity: '58342',
          },
        },
      },
    };
    const batchModulesToUpdate = (x) => [x];

    await updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
      batchModulesToUpdate,
    });
    expect(getModuleMap().getIn(['modules', 'awesome-module'])).toBeUndefined();
  });

  it('should resolve with the new versions of the changed modules', async () => {
    expect.assertions(1);
    const currentModuleMap = getModuleMap().toJS();
    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': {
          node: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.node.js',
            integrity: '4556',
          },
          browser: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.browser.js',
            integrity: '7764',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.legacy.browser.js',
            integrity: '34566',
          },
        },
      },
    };

    const batchModulesToUpdate = (x) => [x];
    const updatedModules = await updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
      batchModulesToUpdate,
    });

    expect(updatedModules).toMatchSnapshot();
  });

  it('should resolve with the new versions of the changed modules if the current module registry is empty', async () => {
    expect.assertions(1);
    // occurs first time updateModuleRegistry is called
    resetModuleRegistry({}, {});

    const currentModuleMap = getModuleMap().toJS();
    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': {
          node: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.node.js',
            integrity: '4556',
          },
          browser: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.browser.js',
            integrity: '7764',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.legacy.browser.js',
            integrity: '34566',
          },
        },
      },
    };

    const batchModulesToUpdate = (x) => [x];
    const updatedModules = await updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
      batchModulesToUpdate,
    });

    expect(updatedModules).toMatchSnapshot();
  });

  it('should throw if any of the modules fail to load', async () => {
    expect.assertions(1);
    const mockLoadModule = async (moduleName, moduleVersion) => `new ${moduleName}@${moduleVersion}`;

    loadModule.mockImplementationOnce(mockLoadModule);
    loadModule.mockImplementationOnce(async () => { throw new Error('Failed to load module'); });
    const currentModuleMap = getModuleMap().toJS();

    const newModuleMap = {
      ...currentModuleMap,
      modules: {
        ...currentModuleMap.modules,
        'another-module': {
          node: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.node.js',
            integrity: '23124',
          },
          browser: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.browser.js',
            integrity: '346346',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/another-module/2.5.6/another-module.legacy.browser.js',
            integrity: '123545',
          },
        },
        'best-module': {
          node: {
            url: 'https://example.com/cdn/best-module/2.5.6/best-module.node.js',
            integrity: '23124',
          },
          browser: {
            url: 'https://example.com/cdn/best-module/2.5.6/best-module.browser.js',
            integrity: '346346',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/best-module/2.5.6/best-module.legacy.browser.js',
            integrity: '123545',
          },
        },
      },
    };
    const batchModulesToUpdate = (x) => x.map((i) => [i]);
    await expect(updateModuleRegistry({
      moduleMap: newModuleMap,
      onModuleLoad,
      batchModulesToUpdate,
    })).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe('areModuleEntriesEqual', () => {
  const moduleEntry = {
    node: {
      url: 'https://example.com/cdn/best-module/2.5.6/best-module.node.js',
      integrity: '23124',
    },
    browser: {
      url: 'https://example.com/cdn/best-module/2.5.6/best-module.browser.js',
      integrity: '346346',
    },
    legacyBrowser: {
      url: 'https://example.com/cdn/best-module/2.5.6/best-module.legacy.browser.js',
      integrity: '123545',
    },
  };

  it('returns true for equal entries', () => {
    expect(areModuleEntriesEqual(moduleEntry, { ...moduleEntry })).toBe(true);
  });

  it('returns false if there is no first module', () => {
    expect(areModuleEntriesEqual(undefined, moduleEntry)).toBe(false);
  });

  it('returns false if there is no second module', () => {
    expect(areModuleEntriesEqual(moduleEntry, undefined)).toBe(false);
  });

  it('returns false if the node url changes', () => {
    const updatedModuleEntry = { ...moduleEntry, node: { ...moduleEntry.node, url: 'changed' } };
    expect(areModuleEntriesEqual(moduleEntry, updatedModuleEntry)).toBe(false);
  });

  it('returns false if the node sri changes', () => {
    const updatedModuleEntry = { ...moduleEntry, node: { ...moduleEntry.node, integrity: 'changed' } };
    expect(areModuleEntriesEqual(moduleEntry, updatedModuleEntry)).toBe(false);
  });

  it('returns false if the browser url changes', () => {
    const updatedModuleEntry = { ...moduleEntry, browser: { ...moduleEntry.browser, url: 'changed' } };
    expect(areModuleEntriesEqual(moduleEntry, updatedModuleEntry)).toBe(false);
  });

  it('returns false if the browser sri changes', () => {
    const updatedModuleEntry = { ...moduleEntry, browser: { ...moduleEntry.browser, integrity: 'changed' } };
    expect(areModuleEntriesEqual(moduleEntry, updatedModuleEntry)).toBe(false);
  });

  it('returns false if the legacyBrowser url changes', () => {
    const updatedModuleEntry = { ...moduleEntry, legacyBrowser: { ...moduleEntry.legacyBrowser, url: 'changed' } };
    expect(areModuleEntriesEqual(moduleEntry, updatedModuleEntry)).toBe(false);
  });

  it('returns false if the legacyBrowser sri changes', () => {
    const updatedModuleEntry = { ...moduleEntry, legacyBrowser: { ...moduleEntry.legacyBrowser, integrity: 'changed' } };
    expect(areModuleEntriesEqual(moduleEntry, updatedModuleEntry)).toBe(false);
  });
});
