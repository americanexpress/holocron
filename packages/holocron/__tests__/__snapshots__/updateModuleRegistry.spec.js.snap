// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateModuleRegistry should remove modules from the registry 1`] = `
Immutable.Map {
  "node": Immutable.Map {
    "url": "https://example.com/cdn/awesome-module/1.0.0/awesome-module.node.js",
    "integrity": "47474",
  },
  "browser": Immutable.Map {
    "url": "https://example.com/cdn/awesome-module/1.0.0/awesome-module.browser.js",
    "integrity": "465345v",
  },
  "legacyBrowser": Immutable.Map {
    "url": "https://example.com/cdn/awesome-module/1.0.0/awesome-module.legacy.browser.js",
    "integrity": "4564632",
  },
}
`;

exports[`updateModuleRegistry should resolve with the new versions of the changed modules 1`] = `
Object {
  "another-module": Object {
    "browser": Object {
      "integrity": "7764",
      "url": "https://example.com/cdn/another-module/2.5.6/another-module.browser.js",
    },
    "legacyBrowser": Object {
      "integrity": "34566",
      "url": "https://example.com/cdn/another-module/2.5.6/another-module.legacy.browser.js",
    },
    "node": Object {
      "integrity": "4556",
      "url": "https://example.com/cdn/another-module/2.5.6/another-module.node.js",
    },
  },
}
`;

exports[`updateModuleRegistry should resolve with the new versions of the changed modules if the current module registry is empty 1`] = `
Object {
  "another-module": Object {
    "browser": Object {
      "integrity": "7764",
      "url": "https://example.com/cdn/another-module/2.5.6/another-module.browser.js",
    },
    "legacyBrowser": Object {
      "integrity": "34566",
      "url": "https://example.com/cdn/another-module/2.5.6/another-module.legacy.browser.js",
    },
    "node": Object {
      "integrity": "4556",
      "url": "https://example.com/cdn/another-module/2.5.6/another-module.node.js",
    },
  },
}
`;

exports[`updateModuleRegistry should throw if any of the modules fail to load 1`] = `"Failed to load module"`;

exports[`updateModuleRegistry should update the modules in the registry: module map 1`] = `
Immutable.Map {
  "key": "key123",
  "modules": Immutable.Map {
    "awesome-module": Immutable.Map {
      "node": Immutable.Map {
        "url": "https://example.com/cdn/awesome-module/1.0.0/awesome-module.node.js",
        "integrity": "47474",
      },
      "browser": Immutable.Map {
        "url": "https://example.com/cdn/awesome-module/1.0.0/awesome-module.browser.js",
        "integrity": "465345v",
      },
      "legacyBrowser": Immutable.Map {
        "url": "https://example.com/cdn/awesome-module/1.0.0/awesome-module.legacy.browser.js",
        "integrity": "4564632",
      },
    },
    "another-module": Immutable.Map {
      "node": Immutable.Map {
        "url": "https://example.com/cdn/another-module/2.4.0/another-module.node.js",
        "integrity": "123",
      },
      "browser": Immutable.Map {
        "url": "https://example.com/cdn/another-module/2.4.0/another-module.browser.js",
        "integrity": "234",
      },
      "legacyBrowser": Immutable.Map {
        "url": "https://example.com/cdn/another-module/2.4.0/another-module.legacy.browser.js",
        "integrity": "344",
      },
    },
    "module-three": Immutable.Map {
      "node": Immutable.Map {
        "url": "https://example.com/cdn/module-three/1.0.0/module-three.node.js",
        "integrity": "7747545",
      },
      "browser": Immutable.Map {
        "url": "https://example.com/cdn/module-three/1.0.0/module-three.browser.js",
        "integrity": "44532345",
      },
      "legacyBrowser": Immutable.Map {
        "url": "https://example.com/cdn/module-three/1.0.0/module-three.legacy.browser.js",
        "integrity": "3534536",
      },
    },
  },
}
`;

exports[`updateModuleRegistry should update the modules in the registry: modules 1`] = `
Object {
  "another-module": Object {
    "$$typeof": Symbol(react.memo),
    "WrappedComponent": [Function],
    "compare": null,
    "displayName": "Connect(HolocronModule(UpdatedModule))",
    "type": [Function],
  },
  "awesome-module": Immutable.Map {
    "$$typeof": Symbol(react.memo),
    "type": [Function],
    "compare": null,
    "WrappedComponent": [Function],
    "displayName": "Connect(HolocronModule(AwesomeModule))",
  },
  "module-three": Immutable.Map {
    "$$typeof": Symbol(react.memo),
    "type": [Function],
    "compare": null,
    "WrappedComponent": [Function],
    "displayName": "Connect(HolocronModule(ModuleThree))",
  },
}
`;
